servers/zms/src/main/java/com/yahoo/athenz/zms/DBService.java

```diff
@@ -6402,6 +6403,147 @@ void enforceGroupUserAuthorityRestrictions(final String domainName, final String
+
+                // Following method does Audit entry as well
+                if (updateGroupMemberDisabledState(null, con, groupMembersWithUpdatedState, currentGroup.getDomainName(),
+                        currentGroup.getGroupName(), ZMSConsts.SYS_AUTH_MONITOR, AUDIT_REF, caller)) {
+                    updatedGroups.add(currentGroup);
+                }
+            }
+            for (GroupMember updatedGM : updatedGroups) {
+
+                // update our group and domain time-stamps, and invalidate local cache entry
+                con.updateGroupModTimestamp(updatedGM.getDomainName(), updatedGM.getGroupName());
+                con.updateDomainModTimestamp(updatedGM.getDomainName());
+                cacheStore.invalidate(updatedGM.getDomainName());
+            }
+        }
+    }
+
+    private void updateRoleMembershipsByPrincipalState(boolean suspended, String caller, ObjectStoreConnection con, Principal updatedUser) {
+        RoleMember roleMember;
+        List<RoleMember> roleMembersWithUpdatedState;
+        DomainRoleMember domainRoleMember;
+        Set<MemberRole> updatedRoles = new HashSet<>();
+
+        domainRoleMember = con.getPrincipalRoles(updatedUser.getFullName(), null);
+        if (!domainRoleMember.getMemberRoles().isEmpty()) {
+            for (MemberRole memberRole : domainRoleMember.getMemberRoles()) {
+                roleMember = new RoleMember();
+                roleMember.setMemberName(updatedUser.getFullName());
+                if (suspended) {
+                    roleMember.setSystemDisabled(memberRole.getSystemDisabled() & ~Principal.State.AUTHORITY_SYSTEM_SUSPENDED.getValue());
+                } else {
+                    roleMember.setSystemDisabled(memberRole.getSystemDisabled() | Principal.State.AUTHORITY_SYSTEM_SUSPENDED.getValue());
+                }
+                roleMembersWithUpdatedState = new ArrayList<>();
+                roleMembersWithUpdatedState.add(roleMember);
+
+                // Following method does Audit entry as well
+                if (updateRoleMemberDisabledState(null, con, roleMembersWithUpdatedState, memberRole.getDomainName(),
+                        memberRole.getRoleName(), ZMSConsts.SYS_AUTH_MONITOR, AUDIT_REF, caller)) {
+                    updatedRoles.add(memberRole);
+                }
+            }
+            for (MemberRole updatedMR : updatedRoles) {
+
+                // update our role and domain time-stamps, and invalidate local cache entry
+                con.updateRoleModTimestamp(updatedMR.getDomainName(), updatedMR.getRoleName());
+                con.updateDomainModTimestamp(updatedMR.getDomainName());
+                cacheStore.invalidate(updatedMR.getDomainName());
+            }
+        }
+    }
+
+    private void updatePrincipalState(boolean suspended, List<Principal> updatedUsers, ObjectStoreConnection con, Principal changedPrincipal) {
+        try {
+            if (suspended) {
```

Comment:
we should simplify this by calling the updatePrincipal only once. 

newState = suspended ? Principal.State.AUTHORITY_SYSTEM_SUSPENDED.getValue() : Principal.State.ACTIVE.getValue();
if (con.updatePrincipal(changedPrincipal.getFullName(), newState)) {
    updatedUsers.add(changedPrincipal);
}